rust/gitxetcore/src/git_integration/git_repo_plumbing.rs

```diff
@@ -357,14 +359,135 @@ pub fn list_files_from_repo(
     Ok(list)
 }
 
+/// Walk the repo working directory starting from search_root.
+/// Return a list of file paths under the search_root, the
+/// file paths are relative to the working dir root.
+/// Note that symlinks are ignored because they are difficult to
+/// deal with: git deals with the symlink file itself without
+/// following the link.
+pub fn walk_working_dir(
+    work_root: impl AsRef<Path>,
+    search_root: impl AsRef<Path>,
+    recursive: bool,
+) -> anyhow::Result<Vec<PathBuf>> {
+    // check existence
+    let work_root = work_root.as_ref();
+    let search_root = search_root.as_ref();
+
+    if !work_root.exists() || !search_root.exists() {
+        return Ok(vec![]);
+    }
+
+    // get absolute paths and check if search below work root
+    let work_root = work_root.canonicalize()?;
+    let search_root = search_root.canonicalize()?;
+
+    if !search_root.starts_with(&work_root) {
+        return Err(anyhow!(
+            "Path {search_root:?} is outside repository at {work_root:?}"
+        ));
+    }
+
+    // ignore .git folder, .gitignore, .gitattributes
+    if is_git_special_files(
+        search_root
+            .file_name()
+            .unwrap_or_default()
+            .to_str()
+            .unwrap_or_default(),
+    ) {
+        return Ok(vec![]);
+    }
+
+    // directly return single file itself
+    if search_root.is_file() {
+        return Ok(vec![search_root.strip_prefix(work_root)?.to_owned()]);
+    }
+
+    // now search_root is a regular directory, walk below it
+
+    let ret: Vec<_> = WalkDir::new(&search_root)
+        .follow_links(false)
+        .max_depth(if recursive { usize::MAX } else { 1 })
+        .into_iter()
+        .filter_entry(|e| !is_git_special_files(e.file_name().to_str().unwrap_or_default()))
```

Comment:
Does this actually exclude the files under .git/ with the walkdir iter now?  It has to be an exact match, so I think the is_git_special_files function needs to test the prefix.

Reply:
Ah, I see that it does.  Nice.